### ISSUE TITLE
Logging structured data without it appearing in the text message

### ISSUE THREAD

**Issue by roji** [aspnet/Extensions#668]
Am looking for some guidance here, or at least confirmation that I understand things right...

Currently, if I want to use the extension methods (e.g. `LogDebug`), it seems that in order to log some piece of structured data it must be referenced from the message - this seems problematic in several scenarios.

First, when using a logging framework such as NLog, it is standard practice to leave the layout of the text message to NLog (user-configurable). Requiring me to embed all structured data in the message text hands off a fully-formed message to NLog, stripping it of one of it's main responsibilities (layouting).

Also, in many scenarios there's a lot of structured context that may be interesting to users in some cases, but that doesn't mean it's *always* interesting and should get included in the text of each and every message. This causes very large log messages with clutter and info nobody really wants (although people may still want the option of looking up these details in the structured data attached to the message).

I understand that I don't have to use the extension methods and can roll my own, but I'd like confirmation to the above, and also your opinion on this.

**Comment by yvdh** [comment]
Hello,
I second the opinion that what's actually inside the message is unrelated to the structured properties you may want to pass on (I use Graylog2 as a logging backend). Also the mandatory quotes for structured string fields in the messages are rather annoying as they cannot be removed (or can they?).
Yves

**Comment by muratg** [comment]
@davidfowl please triage or close this one. Thanks.

**Comment by muratg** [comment]
@ericsampson I don't know.

**Comment by Veikedo** [comment]
Hi all,
Is there any decision on this?

I really liked the syntax @davidfowl proposed.
Here is one of my use cases (simplified):

    public Task SayHello(string clientInfo)
    {
      _logger.LogInformation("SayHello from connection {ConnectionId}", Context.ConnectionId, clientInfo);
      return Task.CompletedTask;
    }

`clientInfo` can be very large and it's not that useful to print it in every log message. However I want to see it in the data bag in elastic.

**Comment by arroyc** [comment]
I have similar use cases as mentioned by @Veikedo .. would love to know if there is any update on this

**Comment by andycmaj** [comment]
IMO it's also not useful to see `SayHello from connection...` in every log message. What you really want to see (and more importantly, track/alert on/graph over time) is the occurrence of `{ event: 'SaidHello', connection: 'MyConnectionId' }`.

**Comment by davidfowl** [comment]
> IMO it's also not useful to see SayHello from connection... in every log message. What you really want to see (and more importantly, track/alert on/graph over time) is the occurrence of { event: 'SaidHello', connection: 'MyConnectionId' }.

You can already log structure, the logger accepts an object. This thread is specifically about text plus other structured data attached to the same log.

This can be implemented today by logging your own objects but you can't use the extension methods in that case. You'd need to create an object that implements `IReadOnlyList<KeyValuePair<string, object>>` and ToString().

https://github.com/aspnet/AspNetCore/blob/6dc65b16730abb9cba59e759012211e576f99188/src/Hosting/Hosting/src/Internal/HostingRequestFinishedLog.cs#L12

https://github.com/aspnet/AspNetCore/blob/f3f9a1cdbcd06b298035b523732b9f45b1408461/src/Hosting/Hosting/src/Internal/HostingApplicationDiagnostics.cs#L190-L195

This call to log specifies the custom object to log, the formatter callback (how to stringify in case the logger doesn't support structure, like the console logger is textual). All of the right pieces are here to accomplish this today we just need to get more crisp on exactly what people need.

**Comment by andycmaj** [comment]
> You can already log structure, the logger accepts an object. This thread is specifically about text plus other structured data attached to the same log.

yes that's way i said, "IMO it's also not useful to see '_SayHello from connection..._' in every log message". Meaning I want to log ONLY the structured data and not grammatical strings of text PLUS a json bag.

I think the real ask is that we have an INTERFACE that encourages and supports pure structured event logging. To me, that means `EventName: string` + `Data: KVP` essentially. `ILogger` extensions are an interface that encourages a formatting string and positional arguments. The example you linked to, 

```
   _logger.Log(
                    logLevel: LogLevel.Information,
                    eventId: LoggerEventIds.RequestFinished,
                    state: new HostingRequestFinishedLog(httpContext, elapsed),
                    exception: null,
                    formatter: HostingRequestFinishedLog.Callback);
``` 

still requires a formatter and isn't really succinct.

I believe the original ask by @roji was whether a PR would be accepted that would add an interface that looked more like:

```
eventLogger.InfoEvent(string eventName, object eventData);
eventLogger.InfoEvent("ProcessedFoos", new { PropA = "bar", PropB = 42 });
```

where the log events just look like

```
{
  "EventName": "ProcessedFoos",
  "Level": "Information",
  "Data": {
    "PropA": "bar",
    "PropB": 42
  }
}
```

**Comment by davidfowl** [comment]
So you're looking for a helper method that does what you can do today to clean up the callsite? That extension method would allocate like crazy (although I'm unsure how you avoid allocations with this design at all).

**Comment by andycmaj** [comment]
yeah agree... but any structured logging that uses anon objects/kvp lists seems like it would have same problem... 

are the allocations for `object eventData` somehow different than a varargs list of format string replacements?

**Comment by davidfowl** [comment]
> yeah agree... but any structured logging that uses anon objects/kvp lists seems like it would have same problem...

That signature you have above while convenient is extremely inefficient and wouldn't be something I'd want in the default package:
- It requires reflection to get properties from the anonymous object (which shouldn't be cached)
- Then another object that allocates IKVP<string, object> needs to be wrapped around that type (it could be a struct though so maybe not the end of the world)
- The object needs to join the event name and the KVPs into a single state object for the ILogger.Log call.

> are the allocations for object eventData somehow different than a varargs list of format string replacements?

The default extension methods have several inefficiencies
- They use a `params object[]` so it always allocates an array if you have a format string, if you don't it's free.
- Being a params object[] means value types are boxed (which results in another allocation)
- Format string need to be parsed on ever call to log (however FormattedLogValues is a struct so it avoids that allocation at least...)

Loggers that implement scopes also need to join the data from the incoming event into and the event itself into an uber collection of properties (more allocations!).

> I believe the original ask by @roji was whether a PR would be accepted that would add an interface that looked more like:

Also I'd like to understand a few things: 
- What happens to loggers that don't support the logging objects and needs to ToString? Seems like you'd still need a formatter to do something reasonable in this case. 
- What does the formatter do for these overloads?
- What happens to exceptions? Do those get appended to the event object as well?
- I see the event object in your sample includes the log level, is that intentional? 

Maybe something like this:

```C#
using System;
using System.Collections.Generic;
using Microsoft.Extensions.Logging;

namespace ConsoleApp68
{
    class Program
    {
        static void Main(string[] args)
        {
            using var loggerFactory = LoggerFactory.Create(builder =>
            {
                builder.AddConsole();
            });

            var logger = loggerFactory.CreateLogger<Program>();

            logger.LogInformation("Hello", new Dictionary<string, object> {
                { "A", 1 }
            });
        }
    }

    public static class LoggerExtensions
    {
        public static void LogInformation<TState>(this ILogger logger, string message, TState state) where TState : IDictionary<string, object>
        {
            state["Message"] = message;
            logger.Log(LogLevel.Information, 0, state, null, (state, exception) => state["Message"].ToString());
        }

        public static void LogInformation<TState>(this ILogger logger, TState state) where TState : IDictionary<string, object>
        {
            logger.Log(LogLevel.Information, 0, state, null, (state, exception) =>
            {
                if (state.TryGetValue("Message", out var value))
                {
                    return (string)value;
                }
                return state.ToString();
            });
        }
    }
}
```

**Comment by jjxtra** [comment]
If they added TState overloads for the Info, Warn, etc. extension methods that would help a lot...

**Comment by TheXenocide** [comment]
We've been migrating from our own logging abstraction to the Microsoft abstraction. We previously implemented our abstraction using NLog (and have tested implementing with ApplicationInsights and verified the same features would be supported with Serilog). We continue to use the NLog implementation of the Microsoft abstraction which is working well for sharing our old analysis tools and we enjoy the flexibility of being able to combine that with console and other loggers (our previous abstraction did not support multiple simultaneous providers, though NLog has support for multiple targets).

Non-message structured data is important to our ability to filter/correlate/analyze/etc. In some cases we use more expensive contextual metadata based on `AsyncLocal<T>` that travels with the call stack which it seems is similar to `BeginScope` in some implementations (and which we have been able to replace/integrate with) but much of our metadata is passed directly to the log method as a dictionary; some are easy enough to convert but in some more crucial cases, as well as in some more abstract reusable utilities we have, we have a need to include the metadata without incorporating it into the message string. On top of that we have scenarios where we want to use both message parameters AND the dictionary and it is unclear how we can possibly do this with the existing infrastructure. It looks like the `TState` approach is more of an either/or scenario where most loggers (at least that matter to us) support the use of a `TState` that implements `IEnumerable<KeyValuePair<string, object>>` but then we can't use message parameters in conjunction. Am I also reading above that in some cases keyed-metadata is ignored unless the key is present in the message string? Ideally we would be looking for something that supports both TState AND message parameters; perhaps along the lines of: `BeingScope<TState>(this ILogger logger, TState state, string message, params object[] args);`, `LogError<TState>(this ILogger logger, EventId eventId, Exception exception, TState state, string message, params object[] args)`, etc. It seems to me like making the FormattedLogValues decorate the state or forcing the state in this case to implement the KVP pattern and concat the values together might be able to make use of the existing allocations/minimize additional impacts, or perhaps an approach that only allocates when the feature is used explicitly which is the developer making a decision to accept the cost in exchange for the benefit.

In the meantime I was thinking maybe we can use a `TState` of `IEnumerable<KeyValuePair<string, object>>` manually including the message parameters as their own keys in the state for structured purposes and then just `String.Format` the values into the message, but then we lose the ability to match on the un-substituted message string which is a handy feature of the message-parameter structured logging approach. This also does not solve our abstract utilities which seek to add metadata to existing message-based logging infrastructure. We really don't want to implement/maintain our own Provider/Factory/Logger, especially when the framework support is so very close to what we (and seemingly many others, especially those using NLog/Serilog/etc.) need and are already used to having in some capacity or another. I'm now taking a look at the above-mentioned `LogValuesCollection` but this really seems like something that could be done in the framework more effectively/efficiently and to the benefit of more people if it were implemented here.

It's also a little unfortunate that in order to determine what types of structured data were actually supported our only option was to literally dig through source code, both here and in the various repositories of the providers. I understand the flexibility of generic state objects for custom structured logging solutions, but it would be really helpful if there were documentation and possibly suggestions for common best practices to urge some level of common ground between implementations (e.g. suggesting that `IEnumerable<KeyValuePair<string, object>>` be supported might help future implementations find common ground). I know this is a challenge and that the API is intentionally flexible, just voicing some of the challenges that were non-trivial in our analysis which could be barriers for adopting common practices in other shops.

**Comment by TheXenocide** [comment]
Ah, it seems that now `FormattedLogValues` and `LogValuesFormatter` are internal which means we can't even benefit from the common code for message-based values in home-rolled structured data types that might aim to use both message parameters and additional state. The `LogValueCollection` mentioned above would have worked for us (though I'd have made a more simple readonly struct based on it), but we're on 3.1 and can no longer do that.

It seems like it would only take a few lines in `FormattedLogValues` if we constrained the state type for additional parameters to `IReadOnlyList<KeyValuePair<string, object>>`. An extra constructor parameter, a couple lines to the indexer and an extra `+ additionalProperties.Count` to the `Count` property? Struct size grows by one pointer no extra allocations (besides whatever the user code does to create their `IReadOnlyList`).

**Comment by TheXenocide** [comment]
Here's what we're prototyping with right now. 

https://gist.github.com/TheXenocide/f7d8a249ec19857dee926426e2319746

Almost all of it is copy/pasted from this repository with a very small number of changes (we can't wait around for .NET 5 to release). If this (or something similar) seems acceptable I might be able to find some time to put together a pull request. Conceptually we wouldn't need the `CustomLogValuesFormatter` at all (it's a straight copy/paste + rename) and we could EITHER keep the `StructuredAndFormattedLogValues` class and use it from the extension methods which need it OR we could just add the functionality into `FormattedLogValues` and only have one class. I can see pros/cons to both. 

Correct me if I'm wrong, but I believe this still satisfies the "No Additional Allocations" goals exactly as well as the current implementation and allows the consumer user to supply an existing Dictionary (which we already have several of in our utilities) or a custom class/struct like the above noted `HostingRequestFinishedLog`. I had considered allowing it to receive `IEnumerable` or `IReadOnlyList` making it either a little more or less flexible but I think this gives a happy middle ground that doesn't require existing dictionaries to be pushed through `ToList`, etc.

**Comment by gitfool** [comment]
We also need this! Using Serilog behind the scenes _but_ using `ILogger` abstraction in libraries and elsewhere to be a good citizen and logger agnostic. Lowest common denominator sucks. 😞 

@TheXenocide please at least try to push for this to be incorporated into .NET 5! 🍺 

Meanwhile, I might take your gist and run with it.

**Comment by snakefoot** [comment]
Since you can provide whatever LogState-objects you like, then it is just a matter a providing one that supports additional properties. 

- [Simple Approach](https://github.com/NLog/NLog.Extensions.Logging/wiki/NLog-properties-with-Microsoft-Extension-Logging#logtstate-simple)

- [Advanced Approach](https://github.com/NLog/NLog.Extensions.Logging/wiki/NLog-properties-with-Microsoft-Extension-Logging#logtstate-advanced) - With message template parsing.

One can sprinkle some ILogger-extension methods around, and then you have what you need.

P.S. Both Serilog and NLog should have no issues with capturing additional properties injected by these two approaches (Important part is that LogState implements `IEnumerable<KeyValuePair<string, object>>`)

**Comment by viktor-nikolaev** [comment]
Just created a nugget package for it 
https://www.nuget.org/packages/XeonApps.Extensions.Logging.WithProperty/

usage:

      // inline
      logger
        .WithProperty("SomeProp", "value")
        .LogInformation("User {User} logged in", "Jon");

      // reassign a logger with props
      logger = logger
        .WithProperty("OneProp", 22)
        .WithProperty("End", 21)
        .WithProperties(
          ("key", "value")
        )
        .WithProperties(
          new KeyValuePair<string, object>("another", "one"),
          new KeyValuePair<string, object>("some", "more"),
          new KeyValuePair<string, object>("End", "more")
        );

      // will have all the added props as well as props from the template 
      logger.LogInformation("Event {Event} happened", "UserLoggedOut");

**Comment by TheXenocide** [comment]
> Since you can provide whatever LogState-objects you like, then it is just a matter a providing one that supports additional properties.
> [...]
> One can sprinkle some ILogger-extension methods around, and then you have what you need.

@snakefoot I think the main difference between the techniques mentioned in your links and a lot of the desire I'm seeing here would be the ability to use both the structured template text AND additional structured properties not included in the message *together*. The examples you linked manage to accomplish logging a structured object which can be turned into a string message, but it doesn't get to leverage the structured template text, thus forcing the user to reinvent the wheel.

I think @viktor-nikolaev manages to accomplish this, though the approach feels a bit more allocation heavy (the example above creates 5 new instances of ILogger to log two messages, though admittedly that's just a verbose sample and it could create 1 new logger per message just fine, which isn't as bad, plus reuse of some properties but not others seems like a viable use of the alternative pattern he provides).

It feels like it would be relatively trivial to modify the provided implementation, and rather non-trivial (and entirely too intimately familiar with the inner workings of the logging abstractions provided here) to provide our own implementation (which also is unable to make any practical reuse of the templating mechanism built-in to the abstraction library, at present). Keeping the standard practice for the structured template strings is really important to us because it is both easy to read and readily referenced all over the internet, rather than us concocting some home-rolled similar-but-different solution everyone will have to relearn and for which there will be few examples of in the wild.

**Comment by snakefoot** [comment]
@TheXenocide Have you tried [Advanced Approach](https://github.com/NLog/NLog.Extensions.Logging/wiki/NLog-properties-with-Microsoft-Extension-Logging#logtstate-advanced) - With message template parsing ? (It does both)
